modules/apps/object/object-service/src/main/java/com/liferay/object/service/impl/ObjectValidationRuleSettingLocalServiceImpl.java

```diff
@@ -53,6 +65,198 @@ public ObjectValidationRuleSetting addObjectValidationRuleSetting(
+			List<ObjectValidationRuleSetting> objectValidationRuleSettings,
+			List<ObjectValidationRuleSetting> oldObjectValidationRuleSettings)
+		throws PortalException {
+
+		Set<String> deleteOldObjectObjectValidationRuleSettingValues =
+			SetUtil.asymmetricDifference(
+				TransformUtil.transform(
+					oldObjectValidationRuleSettings,
+					ObjectValidationRuleSetting::getValue),
+				TransformUtil.transform(
+					objectValidationRuleSettings,
+					ObjectValidationRuleSetting::getValue));
+
+		for (String deleteOldObjectObjectValidationRuleSettingValue :
+				deleteOldObjectObjectValidationRuleSettingValues) {
+
+			ObjectValidationRuleSetting
+				serviceBuilderObjectValidationRuleSetting =
+					objectValidationRuleSettingPersistence.findByOVRI_N_V(
+						objectValidationRuleId, objectValidationRuleSettingName,
+						deleteOldObjectObjectValidationRuleSettingValue);
+
+			deleteObjectValidationRuleSetting(
+				serviceBuilderObjectValidationRuleSetting.
+					getObjectValidationRuleSettingId());
+		}
+	}
+
+	private void _getDeleteObjectValidationRuleSettings(
+		List<ObjectValidationRuleSetting> keyObjectValidationRuleSettings,
+		List<ObjectValidationRuleSetting> objectValidationRuleSettings,
+		List<ObjectValidationRuleSetting> outputObjectValidationRuleSettings) {
+
+		for (ObjectValidationRuleSetting objectValidationRuleSetting :
+				objectValidationRuleSettings) {
+
+			if (objectValidationRuleSetting.compareName(
+					ObjectValidationRuleSettingConstants.
+						NAME_KEY_OBJECT_FIELD_ID)) {
+
+				keyObjectValidationRuleSettings.add(
+					objectValidationRuleSetting);
+
+				continue;
+			}
+
+			outputObjectValidationRuleSettings.add(objectValidationRuleSetting);
+		}
+	}
+
+	private void _validate(
+			ObjectValidationRule objectValidationRule, long userId,
+			List<ObjectValidationRuleSetting> keyObjectValidationRuleSettings,
+			List<ObjectValidationRuleSetting>
+				outputObjectValidationRuleSettings)
+		throws PortalException {
+
+		User user = _userLocalService.getUser(userId);
+
+		if (StringUtil.equals(
+				objectValidationRule.getEngine(),
+				ObjectValidationRuleConstants.ENGINE_TYPE_COMPOSED_KEY) &&
+			(keyObjectValidationRuleSettings.size() < 2)) {
+
+			throw new ObjectValidationRuleSettingCountException(
+				StringBundler.concat(
+					"The ", objectValidationRule.getName(user.getLocale()),
+					" Unique Composed Key Object Validation Rule must have at ",
+					"least two Object Validation Rule Settings to compose the ",
+					"object validation rule"));
+		}
+
+		if (StringUtil.equals(
+				objectValidationRule.getOutputType(),
+				ObjectValidationRuleConstants.OUTPUT_TYPE_PARTIAL_VALIDATION) &&
+			outputObjectValidationRuleSettings.isEmpty()) {
+
+			throw new ObjectValidationRuleSettingCountException(
+				StringBundler.concat(
+					"The ", objectValidationRule.getName(user.getLocale()),
+					" Object Validation Rule must have at least one Object ",
+					"Validation Rule Setting"));
```

Comment:
This same validation is performed [here](https://github.com/liferay/liferay-portal/blob/51b6cbd9e67366bc66456cae09ae678b37e5dc1a/modules/apps/object/object-service/src/main/java/com/liferay/object/service/impl/ObjectValidationRuleLocalServiceImpl.java#L569-L578) too.
Maybe we need to concentrate all _objectValidationRuleSettings_ validation here. For both: the add and the update flow.